File: spring-backend/src/main/java/sg/edu/nus/iss/wellness/error/ApiException.java
package sg.edu.nus.iss.wellness.error;

import org.springframework.http.HttpStatus;

/**
 * API exception carrying an HTTP status code and, optionally, a
 * {@code Retry-After} hint (in seconds) used by throttled {@code 429} responses.
 *
 * @author Tiong Zhong Cheng, Chua Wei Yi Justin
 */
public class ApiException extends RuntimeException {
    private final HttpStatus status;
    private final Long retryAfterSeconds;

    public ApiException(HttpStatus status, String message) {
        this(status, message, null);
    }

    public ApiException(HttpStatus status, String message, Long retryAfterSeconds) {
        super(message);
        this.status = status;
        this.retryAfterSeconds = retryAfterSeconds;
    }

    /** Factory for a throttled response that advertises when the caller may retry. */
    public static ApiException tooManyRequests(String message, long retryAfterSeconds) {
        return new ApiException(HttpStatus.TOO_MANY_REQUESTS, message, retryAfterSeconds);
    }

    public HttpStatus getStatus() {
        return status;
    }

    /** @return seconds until the caller may retry, or {@code null} when not applicable. */
    public Long getRetryAfterSeconds() {
        return retryAfterSeconds;
    }
}

File: dotnet-backend/Errors/ApiException.cs
namespace Wellness.Backup.Api.Errors;

/// <summary>
/// Controlled exception converted to the Spring-compatible JSON error shape.
/// </summary>
/// <remarks>@author Tiong Zhong Cheng, Chua Wei Yi Justin</remarks>
public sealed class ApiException : Exception
{
    public ApiException(int statusCode, string message, long? retryAfterSeconds = null) : base(message)
    {
        StatusCode = statusCode;
        RetryAfterSeconds = retryAfterSeconds;
    }

    public int StatusCode { get; }

    /// <summary>Seconds until the caller may retry, surfaced as a Retry-After header (429 only).</summary>
    public long? RetryAfterSeconds { get; }

    public static ApiException BadRequest(string message) => new(StatusCodes.Status400BadRequest, message);

    public static ApiException TooManyRequests(string message, long retryAfterSeconds) =>
        new(StatusCodes.Status429TooManyRequests, message, retryAfterSeconds);

    public static ApiException Unauthorized(string message) => new(StatusCodes.Status401Unauthorized, message);

    public static ApiException Forbidden(string message) => new(StatusCodes.Status403Forbidden, message);

    public static ApiException NotFound(string message) => new(StatusCodes.Status404NotFound, message);

    public static ApiException Conflict(string message) => new(StatusCodes.Status409Conflict, message);

    public static ApiException ServiceUnavailable(string message) => new(StatusCodes.Status503ServiceUnavailable, message);
}
